File: backend/app/routes/trace.py
"""LLMOps 실행 Trace 조회 API 라우터.

역할: A/C/D가 기록한 실행 단계(step)를 dev-ops 패널에서 볼 수 있도록 집계해 제공한다.
입력: GET /api/trace/stats?since=&until=&recent_errors_limit=
출력: TraceStatsResponse.
호출 시점: GET stats는 dev-ops 패널(TracePanel, TP-157)이 호출한다.
      세션 단위 조회(get_traces)는 별도 API로 노출하지 않는다 — 현재
      필요한 것은 "step별 평균 지연시간", "최근 에러"처럼 세션을 가리지
      않는 통계뿐이다(agent-state-contract-v1.md 경계 원칙: B는 세션 내부
      trace의 의미를 해석하지 않고 저장·집계만 한다).
"""

from __future__ import annotations

from datetime import datetime

from fastapi import APIRouter

from app.auth.dependency import OptionalPrincipal
from app.state import service as state_service

router = APIRouter(tags=["trace"])


@router.get("/trace/stats", response_model=state_service.TraceStatsResponse)
async def get_trace_stats(
    principal: OptionalPrincipal,
    since: datetime | None = None,
    until: datetime | None = None,
    recent_errors_limit: int = 20,
) -> state_service.TraceStatsResponse:
    return state_service.get_trace_stats(
        since=since, until=until, recent_errors_limit=recent_errors_limit
    )

File: frontend/src/api/trace.ts
/*
 * 역할: LLMOps 실행 Trace 통계 API 함수.
 * 입력: 통계 조회 시 since·until·recentErrorsLimit.
 * 출력: 백엔드가 반환한 step별 집계 + 최근 에러 목록.
 * 호출 시점: dev-ops 패널(TracePanel, TP-157)이 호출한다 — 이 엔드포인트는
 *      dev.ts의 다른 함수들과 달리 APP_ENV=local 여부와 무관하게 항상 등록돼 있다
 *      (backend/app/main.py, trace_router는 무조건 include).
 * 참고: backend/app/routes/trace.py GET /api/trace/stats.
 */

import { apiClient } from "./client";
import type { TraceStatsResponse } from "../types";

export function fetchTraceStats(params?: {
  since?: string;
  until?: string;
  recentErrorsLimit?: number;
}) {
  const query = new URLSearchParams();
  if (params?.since) query.set("since", params.since);
  if (params?.until) query.set("until", params.until);
  if (params?.recentErrorsLimit !== undefined) {
    query.set("recent_errors_limit", String(params.recentErrorsLimit));
  }
  const qs = query.toString();
  return apiClient.get<TraceStatsResponse>(`/trace/stats${qs ? `?${qs}` : ""}`);
}
